q2_types/_feature_data.py

```diff
@@ -0,0 +1,183 @@
+import pandas as pd
+
+import qiime
+from qiime.plugin import SemanticType, DataLayout, FileFormat
+from .plugin_setup import plugin
+
+
+FeatureData = SemanticType('FeatureData', field_names='type')
+
+Taxonomy = SemanticType('Taxonomy', variant_of=FeatureData.field['type'])
+
+Sequence = SemanticType('Sequence', variant_of=FeatureData.field['type'])
+
+AlignedSequence = SemanticType('AlignedSequence',
+                               variant_of=FeatureData.field['type'])
+
+
+class TaxonomyFormat(FileFormat):
+    name = 'taxonomy'
+
+    @classmethod
+    def sniff(cls, filepath):
+        with open(filepath, 'r') as fh:
+            count = 0
+            while count < 10:
+                line = fh.readline()
+                if line == '':
+                    break
+                elif line.startswith('#'):
+                    continue
+                else:
+                    cells = line.split('\t')
+                    if len(cells) < 2:
+                        return False
+                    count += 1
+            return False if count == 0 else True
+
+
+class FASTAFormat(FileFormat):
+    name = 'fasta'
+
+    @classmethod
+    def sniff(cls, filepath):
+        sniffer = skbio.io.io_registry.get_sniffer('fasta')
+        return sniffer(filepath)[0]
+
+
+class AlignedDNAFASTAFormat(FileFormat):
+    name = 'aligned-dna-fasta'
+
+    @classmethod
+    def sniff(cls, filepath):
+        sniffer = skbio.io.io_registry.get_sniffer('fasta')
+        if sniffer(filepath)[0]:
+            generator = skbio.io.read(filepath, constructor=skbio.DNA,
+                                      format='fasta')
+            try:
+                initial_length = len(next(generator))
+                for seq, _ in zip(generator, range(4)):
+                    if len(seq) != initial_length:
+                        return False
+                return True
+            # ValueError raised by skbio if there are invalid DNA chars.
+            except (StopIteration, ValueError):
+                return False
+        else:
+            return False
+
+
+taxonomy_data_layout = DataLayout('taxonomy', 1)
+taxonomy_data_layout.register_file('taxonomy.tsv', TaxonomyFormat)
+
+sequences_data_layout = DataLayout('sequences', 1)
+sequences_data_layout.register_file('sequences.fasta', FASTAFormat)
+
+aligned_dna_sequences_data_layout = DataLayout('aligned-dna-sequences', 1)
+aligned_dna_sequences_data_layout.register_file('aligned-dna-sequences.fasta',
+                                                AlignedDNAFASTAFormat)
+
+
+def taxonomy_to_pandas_series(data_dir):
+    filepath = os.path.join(data_dir, 'taxonomy.tsv')
+    # Using read_csv for access to comment parameter, and set index_col,
+    # header, and parse_dates to Series.from_csv default settings for
+    # better handling of round-trips
+    df = pd.read_csv(filepath, sep='\t', comment='#', index_col=0, header=0,
+                     parse_dates=True, skip_blank_lines=True)
```

Comment:
In the past what we've done is parse like you did except we don't
specify the index column. Once the file is parsed and in the dataframe
we set a particular column as the index and 🎉.

On (Jul-22-16| 8:56), Jai Ram Rideout wrote:
> > +
> > +sequences_data_layout = DataLayout('sequences', 1)
> > +sequences_data_layout.register_file('sequences.fasta', FASTAFormat)
> > +
> > +aligned_dna_sequences_data_layout = DataLayout('aligned-dna-sequences', 1)
> > +aligned_dna_sequences_data_layout.register_file('aligned-dna-sequences.fasta',
> > +                                                AlignedDNAFASTAFormat)
> > +
> > +
> > +def taxonomy_to_pandas_series(data_dir):
> > +    filepath = os.path.join(data_dir, 'taxonomy.tsv')
> > +    # Using read_csv for access to comment parameter, and set index_col,
> > +    # header, and parse_dates to Series.from_csv default settings for
> > +    # better handling of round-trips
> > +    df = pd.read_csv(filepath, sep='\t', comment='#', index_col=0, header=0,
> > +                     parse_dates=True, skip_blank_lines=True)
> 
> @ElDeveloper do you know how to work around that issue? AFAICT there isn't an option in `pd.read_csv` to specify the index dtype.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/qiime2/q2-types/pull/18/files/7c176a117e70a99990402b128295f304c0b9a763#r71902057